walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,85 @@
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> empty = new ArrayList<>();
+    ArrayList<String> mandatoryAttendees = new ArrayList<String>(request.getAttendees()); 
+    ArrayList<String> allAttendees = new ArrayList<String>(mandatoryAttendees);
+    allAttendees.addAll(request.getOptionalAttendees());
+    // return whole day if there are no attendees
+    if (allAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } 
+    //return empty If there are no events during day(gaps), but there are optional attendees
+    if(events.isEmpty() && !request.getOptionalAttendees().isEmpty()){
+      return empty;
+    }
+    // When there are no available times with optional attendees, set available times for mandatory employees only
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>(
+        availableTimesForAttendees(events, allAttendees, request.getDuration()));
+    if(!availableTimes.isEmpty()){
+      return availableTimes;
+    }else{
+      // If the request has no attendees after optional attendees are excluded, return an empty collection.
+      if(!mandatoryAttendees.isEmpty()){
+        availableTimes = new ArrayList<TimeRange>(availableTimesForAttendees(events, mandatoryAttendees, request.getDuration()));
+        return availableTimes;
+      }else{
+        return empty;
+      }
+    }
+  }
+  /**
+   * @param events a collection of events going on during the day
+   * @param attendees a list that contains the attendees for the available times being searched
+   * @param duration length of the user requested meeting
+   * @return a list that holds the available times to schedule a meeting for the current atendees
+   */
+  public Collection<TimeRange> availableTimesForAttendees(Collection<Event> events, ArrayList<String> attendees, long duration) {
+    ArrayList<TimeRange> unavailableTimes = new ArrayList<>();
+
+    //Find unavailable times for meetings
+    for(Event event: events){
+      if(!Collections.disjoint(event.getAttendees(), attendees)){
+        unavailableTimes.add(event.getWhen());
+      }
+    }
+    //Sort the list of unavailable times in ascending order
+    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+
+    //Find available times for meetings
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    int meetingStartTime = TimeRange.START_OF_DAY;
+    for(TimeRange time: unavailableTimes){
+      if(time.contains(meetingStartTime)){
+        meetingStartTime = Math.max(time.end(),meetingStartTime);
+      }else{
+        if((time.start() - meetingStartTime) >= duration){
+          availableTimes.add(TimeRange.fromStartEnd(meetingStartTime, time.start(), false));
+        }
+        if(meetingStartTime > time.end()){
+          continue;
+        }
+        meetingStartTime = time.end();
```

Comment:
do 
if(meetingStartTime < time.end()) {
   meetingStartTime = time.end();
}
here instead. it's a little more readable